### PR TITLE
chore(ls): Make semgrep lsp.js easy to distribute

### PR DESCRIPTION
**Pull request by ajbt200128**
This PR makes it so we don't have to distribute a ton of parser files w/ lsp.js when we want to distribute it.

This happens by:
* Building lsp.js with the parsing ocaml code (instead of as a separate module)
* Compiling the C code of parsers/tree sitter to wasm objects, instead of js/wasm files
* Using said objects to make one wasm file for lsp.js instead of 1 for each language
* Some refactoring

now there are only 4 files in js/language_server/dist that we need to distribute for lsp.js! Easy.

## Test plan
Tests in js/language_server pass

### CODE REVIEW

File: js/language_server/Makefile [hunk 1]
```diff
@@ -2,9 +2,19 @@
...
+include ../tree-sitter/Makefile.include
+
+# TODO: manage this better
+LANGUAGES := bash c cairo cpp csharp dart dockerfile elixir go hack html java jsonnet kotlin lisp lua ocaml php promql protobuf python r rust solidity terraform typescript vue
+TS_LANGS := bash c cairo clojure cpp c_sharp dart dockerfile elixir go hack html java jsonnet kotlin lua ocaml php promql proto python r rust solidity hcl typescript vue
+
+empty:=
+space:= $(empty) $(empty)
+comma:=,
+PARSER_EXPORTED_FUNCTIONS := $(subst $(space),$(comma),$(TS_LANGS:%=_tree_sitter_%))
```
Comment (tpetr) [inline js/language_server/Makefile]: _cries in Makefile_

File: js/language_server/Makefile [hunk 2]
```diff
@@ -23,11 +33,25 @@ package-lock.json:
 node_modules/.package-lock.json: package-lock.json
 	npm ci
 
+../tree-sitter/dist/tree-sitter.o:
+	$(MAKE) -C ../tree-sitter dist/tree-sitter.o
+
+dist/language-server-c-bindings.js: ../libyaml/dist/libyaml.o ../libpcre/dist/libpcre.a ../tree-sitter/dist/tree-sitter.o $(LANGUAGES:%=../languages/%/dist/semgrep-parser.o)
```
Comment (tpetr) [inline js/language_server/Makefile]: nit: the naming (`c-bindings`) feels confusing to me, thoughts on `semgrep-language-server-wasm` or `semgrep-ls-wasm`?

File: js/language_server/Makefile (hunk 1 above)
Comment (ajbt200128) [reply js/language_server/Makefile]: I want to replace this w/ a sed command but I'm too lazy

File: js/language_server/Makefile [hunk 3]
```diff
@@ -23,11 +33,25 @@ package-lock.json:
...
+	mkdir -p dist
+	emcc \
+		-O3 \
+		$^ \
+		$(EMCC_DEFAULTS) \
+		-s INITIAL_MEMORY=31457280 \
+		-s SINGLE_FILE=1 \
+		-sEXPORTED_FUNCTIONS=_malloc,_free,$(YAML_EXPORTED_FUNCTIONS),$(PCRE_EXPORTED_FUNCTIONS),$(TREESITTER_EXPORTED_FUNCTIONS),$(PARSER_EXPORTED_FUNCTIONS) \
+		-o $(word 1,$@)
```
Comment (tpetr) [inline js/language_server/Makefile]: you can replace `$(word 1,$@)` with `$@` -- the `word` was needed when this had multiple targets

File: js/language_server/dune [hunk 4]
```diff
@@ -21,6 +20,14 @@
...
  ))
  (preprocess
   (pps lwt_ppx js_of_ocaml-ppx)))
+
+(rule
+ (targets semgrep.js)
+ (action
+  (with-stdout-to
+   %{targets}
+   (run ../languages/shared/generate-semgrep-js.sh bash c cairo cpp c_sharp dart dockerfile elixir go hack html java json jsonnet julia kotlin lisp lua ocaml php promql proto python r ruby rust scala solidity swift terraform typescript vue tsx hcl clojure))))
```
Comment (tpetr) [inline js/language_server/dune:34]: can you add a TODO comment to clean this up?

File: js/languages/dockerfile/Makefile [hunk 5]
```diff
@@ -1,7 +1,8 @@
 SEMGREP_LANG := dockerfile
 
 # dockerfile parser also needs bash parser
-EXTRA_TREE_SITTER_SOURCES := $(shell find ../../../languages/bash/tree-sitter -name 'scanner.c' -o -name 'scanner.cc' -o -name 'parser.c')
+#EXTRA_TREE_SITTER_SOURCES := $(shell find ../../../languages/bash/tree-sitter -name 'scanner.c' -o -name 'scanner.cc' -o -name 'parser.c')
```
Comment (tpetr) [inline js/languages/dockerfile/Makefile]: this can be deleted

File: js/language_server/src/semgrep-lsp-bindings.ts [hunk 6]
```diff
@@ -1,50 +1,4 @@
...
-    parserInstance
-      .getLangs()
-      .forEach((lang: string) => parsers.set(lang, parserInstance));
-  };
-  for (const parserName of languages) {
-    await addParser(parserName);
-  }
-  return parsers;
-};
+const LSWasm = require("./language-server-c-bindings");
```
Comment (tpetr) [inline js/language_server/src/semgrep-lsp-bindings.ts]: nit: can we massage the names to match a little better here?

**Review by tpetr — APPROVED** [review]
ez

File: js/language_server/Makefile [hunk 7]
```diff
@@ -2,9 +2,19 @@
 default: build
 
 include ../engine/Makefile.include
+include ../tree-sitter/Makefile.include
+
+# TODO: manage this better
+LANGUAGES := bash c cairo cpp csharp dart dockerfile elixir go hack html java jsonnet kotlin lisp lua ocaml php promql protobuf python r rust solidity terraform typescript vue
+TS_LANGS := bash c cairo clojure cpp c_sharp dart dockerfile elixir go hack html java jsonnet kotlin lua ocaml php promql proto python r rust solidity hcl typescript vue
```
Comment (tpetr) [inline js/language_server/Makefile]: could use a comment explaining what the differences between `LANGUAGES` and `TS_LANGS` is

File: js/language_server/Makefile (hunk 2 above)
Comment (ajbt200128) [reply js/language_server/Makefile]: yea sure

**Comment by ajbt200128** [comment]
PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

File: js/language_server/Makefile [hunk 8]
```diff
@@ -2,9 +2,23 @@
...
+# All languages that have a C parser we need to compile w/emcc
+LANGUAGES := bash c cairo cpp csharp dart dockerfile go hack html java jsonnet kotlin lisp lua ocaml php promql protobuf python r rust solidity terraform typescript vue
+
+# Tree Sitter parser names so we can properly export them in EMCC
+TS_LANGS := bash c cairo clojure cpp c_sharp dart dockerfile go hack html java jsonnet kotlin lua ocaml php promql proto python r rust solidity hcl typescript vue
+
+empty:=
+space:= $(empty) $(empty)
+comma:=,
+PARSER_EXPORTED_FUNCTIONS := $(subst $(space),$(comma),$(TS_LANGS:%=_tree_sitter_%))
```
Comment (aryx) [inline js/language_server/Makefile]: please use sed or something else. This is ugly.

**Comment by aryx** [comment]
Please find a way to include your PR description somewhere in the code or Makefile. The general organization principle is useful to have
in the Makefile.

File: js/language_server/Makefile (hunk 8 above)
Comment (ajbt200128) [reply js/language_server/Makefile]: So fair

**Review by aryx — APPROVED** [review]
Thx for the README and the sed :)

**Pull request merged**